### PR TITLE
[diffusion] Fix Z-Image accuracy

### LINKED ISSUE #28502: Dynamic batching causes severe image quality degradation, line art distorted and subjects missing with same seed & steps
### Checklist

- [x] I searched related issues but found no solution.
- [ ] The bug persists in the latest version.
- [ ] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

Hello everyone, I would like to ask my friends a question. After setting up the ` -- bathing-mode dynamic `, I turned on the online inference service. Then two scripts were written, one is the concurrent stress testing script batching_test.py, which is used to send batch requests to the SGLang graph service, simulating multiple people sending requests to the server. Another script, openai_test. py, uses the OpenAi interface to send a single request to the SGLang graph service. However, for the same prompt words, the output images are not the same, and the images generated by batching_test. py for concurrent stress testing scripts have a high probability of distorted lines or missing main characters (which is very common when the prompt words are in Chinese). The image generated by calling the OpenAi interface script openai_test. py will be much better, very fitting the prompt words. I also found that as the number of requests increases, the errors become more severe. I checked the seed and infer_steps, and they are all consistent. To prevent uncertainty, I turned off Cache DiT acceleration. And when I set ` -- batch max size 1 `, the generated image effect will be significantly better, and the result will be the same as the image generated by a single call to the OpenAI interface.
I have reviewed the document and found the document link https://docs.sglang.io/docs/sglang-diffusion/dynamic_batching#notes mention:
~~~text
Batch shape can change kernels, so singleton and dynamic outputs are not expected to be bit-exact.
~~~
I understand from the docs that "singleton and dynamic outputs are not expected to be bit-exact", but the difference here is not minor pixel-level variation — the generated image is severely distorted with missing subjects, which is far from acceptable generation quality.
Is there any way to solve the problem of poor image generation requests when batch requests are made after setting -- batch mode dynamic?
The same prompt word, the image generated by OpenAI interface and the image generated by test. py are as follows:

Images generated by OpenAI interface  ` openai_test.py `   :
image
image

The image generated by ` batching_test. py `
image

image

### Reproduction

The following are reproducible methods:
Command to start high concurrency service:
~~~python
#!/bin/bash
sglang serve \
   --model-path /root/autodl-fs/data/z-image_turbo/Tongyi-MAI/Z-Image-Turbo \
   --host 0.0.0.0 \
   --port 30000 \
   --batching-mode dynamic \
   --batching-max-size 5 \
   --batching-delay-ms 50 \
   --enable-batching-metrics \
   2>&1 | tee sglang_sana_q.log
~~~

The code for OpenAi interface ` openai_test.py `:
~~~python
import base64
from openai import OpenAI

with open("rule4.txt",encoding="utf-8") as f:
    rule_data=f.read().strip()
prompt="Naruto and Luffy sit side by side on a large boulder. On the left, Naruto has a smile on his face and raises one hand to make a victory gesture. On the right, Luffy wears a wide-brimmed straw hat, an unbuttoned short-sleeved jacket, shorts and flip-flops, with a long sword strapped to his back. He is also grinning happily. Behind them lie a wide expanse of water and the sky dotted with a few clouds, and patches of grass grow beside the boulder."

client = OpenAI(api_key="EMPTY", base_url="http://localhost:30000/v1")
img = client.images.generate(
    model="Tongyi-MAI/Z-Image-Turbo",
    # prompt=prompt,
    prompt=rule_data+prompt,
    size="640x480", 
    n=1,
    response_format="b64_json",
    extra_body={"seed": 42,"num_inference_steps":9} 
)
name=prompt[:10]
image_bytes = base64.b64decode(img.data[0].b64_json)
with open("./{}1.png".format(name), "wb") as f:
    f.write(image_bytes)
~~~

Test concurrent code ` batching_test. py `:
~~~python
import asyncio
import aiohttp
import base64
import os
import random

API_URL = "http://localhost:30000/v1/images/generations"
MODEL_NAME = "Tongyi-MAI/Z-Image-Turbo"
TOTAL_REQUESTS = 6
BATCH_SIZE = 5

SAVE_IMAGES = True
SAVE_COUNT = TOTAL_REQUESTS

OUTPUT_DIR = "imgs"
DATA_FILE = "data4.txt"
os.makedirs(OUTPUT_DIR, exist_ok=True)
success_count = 0
fail_count = 0
with open("rule4.txt",encoding="utf-8") as f:
    rule_data=f.read().strip()
def load_prompts(file_path, count):
    with open(file_path, 'r', encoding='utf-8') as f:
        prompts = [line.strip() for line in f if line.strip()]
    if not prompts:
        raise ValueError("Prompt file is empty")
    if len(prompts) >= count:
        return random.sample(prompts, count)
    print(f"[Warning] Prompt count ({len(prompts)}) is less than request count ({count}), prompts will be reused.")
    original = prompts[:]
    while len(prompts) < count:
        prompts.append(random.choice(original))
    return prompts

async def generate_image(session, request_id, prompt, semaphore):
    global success_count
    global fail_count
    async with semaphore:
        payload = {
            "model": MODEL_NAME,
            "prompt": rule_data+prompt,
            # "prompt": prompt,
            "n": 1,
            "size": "640x480",
            "response_format": "b64_json",
            "seed":42,
            "num_inference_steps":9
        }
        try:
            async with session.post(API_URL, json=payload, timeout=300) as response:
                if response.status != 200:
                    fail_count += 1
                    print(f"[Failed] Request {request_id} HTTP {response.status}")
                    return
                result = await response.json()
                success_count += 1
                if SAVE_IMAGES and request_id < SAVE_COUNT:
                    image_b64 = result["data"][0]["b64_json"]
                    image_bytes = base64.b64decode(image_b64)
                    prompt_prefix = prompt[:9]
                    with open(f"{OUTPUT_DIR}/{prompt_prefix}.png", "wb") as f:
                        f.write(image_bytes)
        except Exception as e:
            fail_count += 1
            print(f"[Error] Request {request_id}: {e}")


async def main():
    prompts = load_prompts(DATA_FILE, TOTAL_REQUESTS)
    semaphore = asyncio.Semaphore(BATCH_SIZE)
    connector = aiohttp.TCPConnector(limit=0)
    async with aiohttp.ClientSession(connector=connector) as session:

        total_batches = (TOTAL_REQUESTS + BATCH_SIZE - 1) // BATCH_SIZE
        request_counter = 0
        for batch_id in range(total_batches):
            batch_prompts = prompts[
                batch_id * BATCH_SIZE:
                (batch_id + 1) * BATCH_SIZE
            ]
            tasks = []
            for prompt in batch_prompts:
                tasks.append(
                    asyncio.create_task(
                        generate_image(
                            session,
                            request_counter,
                            prompt,
                            semaphore
                        )
                    )
                )
                request_counter += 1
            await asyncio.gather(*tasks)
            print(
                f"[Batch {batch_id+1}] "
                f"Success={success_count} Failed={fail_count}"
            )
    print("\n========== SUMMARY ==========")
    print(f"Success: {success_count}")
    print(f"Failed: {fail_count}")

if __name__ == "__main__":
    asyncio.run(main())
~~~
Prompt word data4.txt:
~~~text
Draw the Titanic.
Draw an illustration of Ultraman fighting a monster.
Draw Donald Duck and Mickey Mouse fishing by the river.
Crayon Shin chan walks with a puppy on the street.
Naruto and Luffy sit side by side on a large boulder. On the left, Naruto has a smile on his face and raises one hand to make a victory gesture. On the right, Luffy wears a wide-brimmed straw hat, an unbuttoned short-sleeved jacket, shorts and flip-flops, with a long sword strapped to his back. He is also grinning happily. Behind them lie a wide expanse of water and the sky dotted with a few clouds, and patches of grass grow beside the boulder.
Draw a banana.
~~~
Preset prompt word rule4.txt
~~~text
Please produce a black-and-white line art drawing.

**Specifications**: Render the artwork exclusively with black lines, with distinct, well-defined outlines. The piece shall consist solely of black linework on a white background, with no fills, no gradients, and no shading.
Emphasize clarity of outlines and structural forms. All lines must be sharp, continuous, and with clean edges; the overall image shall be neat and uncluttered.

The subject to be depicted is:
~~~

### Environment

(sglang-env) root@autodl-container-uudafed8ye-7b8430ad:~# python3 -m sglang.check_env
Python: 3.12.3 | packaged by Anaconda, Inc. | (main, May  6 2024, 19:46:43) [GCC 11.2.0]
CUDA available: True
GPU 0: NVIDIA GeForce RTX 4090
GPU 0 Compute Capability: 8.9
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.8, V12.8.93
CUDA Driver Version: 580.105.08
PyTorch: 2.11.0+cu130
sglang: 0.5.12.post1
sglang-kernel: 0.4.2.post2
flashinfer_python: 0.6.11.post1
flashinfer_cubin: 0.6.11.post1
flashinfer_jit_cache: Module Not Found
triton: 3.6.0
transformers: 5.6.0
torchao: 0.17.0
numpy: 2.3.5
aiohttp: 3.14.0
fastapi: 0.136.3
huggingface_hub: 1.17.0
interegular: 0.3.3
modelscope: 1.37.1
orjson: 3.11.9
outlines: 0.1.11
packaging: 26.2
psutil: 7.2.2
pydantic: 2.14.0a1
python-multipart: 0.0.30
pyzmq: 27.1.0
uvicorn: 0.48.0
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.2.0
openai: 2.6.1
tiktoken: 0.13.0
anthropic: 0.105.2
litellm: Module Not Found
torchcodec: 0.11.1
NVIDIA Topology: 
        GPU0    NIC0    NIC1    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      SYS     SYS     32-63,96-127    1               N/A
NIC0    SYS      X      PIX
NIC1    SYS     PIX      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1

### PR DESCRIPTION
**Pull request by qimcis**
## Motivation

Fix #28502 , and general accuracy issues with Z-image-Turbo

## Modifications

With `--batching-mode dynamic`, `Tongyi-MAI/Z-Image-Turbo` could produce severely degraded images when requests were batched. 

Initially, i found part of this issue was that qwen3 text encoder default `position_ids` were shaped `[1, seq]` even when `hidden_states` were `[batch, seq, dim]`, causing RoPE to be applied with the wrong flattened token layout for batched prompts,

But following this, I realize that our current z-image turbo implementation is severely degraded, even for singleton generations, compared to the [native pytorch implementation](https://github.com/Tongyi-MAI/Z-Image):

Firstly, z-image sampling and normalization differed from the native implementation above; we use the scheduler default sigma path, outer autocast, and shared fp32-accumulating RMSNorm,  changing the denoising trajectory and bf16 activations for this model

Secondly, we padded batched images/captions to shared batch len but did not preserve RoPE offsets for each request or mask extra batch-only padding in attention, which causes mixed-prompt dynamic batches to attend to invalid tokens and use wrong image positions

## Accuracy Tests

Tested on 1xh100

```
python3 -m sglang.multimodal_gen.runtime.entrypoints.cli.main serve \
  --model-path Tongyi-MAI/Z-Image-Turbo \
  --host 0.0.0.0 \
  --port 30000 \
  --batching-mode dynamic \
  --batching-max-size 5 \
  --batching-delay-ms 1000 \
  --enable-batching-metrics
```
```
{
  "model": "Tongyi-MAI/Z-Image-Turbo",
  "prompt": "Please produce a black-and-white line art drawing.\n\nSpecifications: Render the artwork exclusively with black lines, with distinct, well-defined outlines. The piece shall consist solely of black linework on a white background, with no fills, no gradients, and no shading. Emphasize clarity of outlines and structural forms. All lines must be sharp, continuous, and with clean edges; the overall image shall be neat and uncluttered.\n\nThe subject to be depicted is: <PROMPT>",
  "n": 1,
  "size": "640x480",
  "response_format": "b64_json",
  "seed": 42,
  "num_inference_steps": 8,
  "guidance_scale": 0.0
}
```
Prompts tested:

- `Draw Donald Duck and Mickey Mouse fishing by the river.`
- `Crayon Shin chan walks with a puppy on the street.`
- `Naruto and Luffy sit side by side on a large boulder. On the left, Naruto has a smile on his face and raises one hand to make a victory gesture. On the right, Luffy wears a wide-brimmed straw hat, an unbuttoned short-sleeved jacket, shorts and flip-flops, with a long sword strapped to his back. He is also grinning happily. Behind them lie a wide expanse of water and the sky dotted with a few clouds, and patches of grass grow beside the boulder.`
- `SpongeBob SquarePants and Patrick sat at the table having dinner together, and there was a TV in the middle of the house with an old phone next to it.`
- `Mario is driving a go kart on the highway, surrounded by many low trees.`

all_prompts_native_main_branch_singleton_batch5_no_prompt_column

## Speed Tests and Profiling

1xh100, 1024x1024, 9 steps single generation
| | Loop Time |
|---|---:|
| main | `863.27 ms` |
| current branch | `851.57 ms` |

batch generation size 5:
|  | Batch Wall | Mean Client | Denoising | Avg Step | 
|---|---:|---:|---:|---:|
| main | `4907.5 ms` | `4893.5 ms` | `3749.0 ms` | `416.3 ms` | 
| current branch | `4781.3 ms` | `4765.0 ms` | `3691.5 ms` | `409.9 ms` | 

about 2% faster across single and batch generation

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #28860860153](https://github.com/sgl-project/sglang/actions/runs/28860860153)
Latest PR Test (Extra): :x: [Run #28860859883](https://github.com/sgl-project/sglang/actions/runs/28860859883)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/models/dits/zimage.py [hunk 1]
```diff
@@ -884,15 +934,139 @@ def patchify_and_embed(
...
+        device: torch.device,
+    ) -> Tuple[Optional[torch.Tensor], Optional[dict]]:
+        if all(length == image_target_len for length in image_lengths) and all(
+            length == cap_target_len for length in cap_lengths
+        ):
+            return None, None
+
+        image_pos = torch.arange(image_target_len, device=device).unsqueeze(0)
+        cap_pos = torch.arange(cap_target_len, device=device).unsqueeze(0)
+        image_len = torch.tensor(image_lengths, device=device).unsqueeze(1)
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/dits/zimage.py]: nit: `torch.as_tensor`

**Review by mickqian — COMMENTED** [review]
excellent. could you help confirm:
1. this kind of problem only happens with Z-Image, and,
2. does this [ground truth](https://github.com/sgl-project/ci-data/blob/main/diffusion-ci/consistency_gt/official_generated/zimage_image_t2i_1gpu.jpg) needs to be updated, or we just need to tighten the consistency threshold? repro script is [here](https://github.com/sgl-project/ci-data/blob/main/diffusion-ci/repro_scripts/gen_official_diffusion_gt.py)

**Comment by qimcis** [comment]
> excellent. could you help confirm:
> 
> 1. this kind of problem only happens with Z-Image, and,
> 2. does this [ground truth](https://github.com/sgl-project/ci-data/blob/main/diffusion-ci/consistency_gt/official_generated/zimage_image_t2i_1gpu.jpg) needs to be updated, or we just need to tighten the consistency threshold? repro script is [here](https://github.com/sgl-project/ci-data/blob/main/diffusion-ci/repro_scripts/gen_official_diffusion_gt.py)

i tested across some other models, and didn't seem to see the same problem - gt also seems to be okay and i don't think it needs to be updated, let's see if ci passes

**Comment by mickqian** [comment]
could you resolve the conflict? cheers

**Review by mickqian — APPROVED** [review]
